update proxy blog post (#777)
linkerd/website: linkerd.io/content/blog/under-the-hood-of-linkerd-s-state-of-the-art-rust-proxy-linkerd2-proxy.md

rather than sticking to Go or Scala, or some more "traditional" proxy language
like C++ or C?

The decision to use Rust came down to several factors. First, a service mesh
proxy has some pretty stringent requirements: because it's deployed as a
sidecar on a per-pod basis, it has to have as small a memory and CPU footprint
as possible. Because most or all of the application's network traffic flows
through the proxy, it needs to have minimal latency overhead, [especially
worst-case tail
latency](https://cacm.acm.org/magazines/2013/2/160173-the-tail-at-scale/fulltext).
Perhaps most importantly, because the proxy handles _application_
data—potentially including incredibly sensitive data such as financial
transitionacitons or personal health—it has to be _secure_.

Let's take each of these in turn, starting with resource consumption. Before we
wrote Linkerd2-proxy, we built Linkerd 1.x. This first version of Linkerd had a
proxy component that was written in Scala, and leveraged the robust Scala and
Java networking ecosystem for excellent performance at scale. But, because it
ran on the Java Virtual Machine, it had a pretty significant resource

> footprint. (The JVM is good at "scaling up", but not good at "scaling _down"_,

Reviewer: i think the quote should not be within the italic part
```suggestion
footprint. (The JVM is good at "scaling up", but not good at "scaling _down_",
```

as William wrote in his InfoQ article on [the decision to reimplement
Linkerd](https://www.infoq.com/articles/linkerd-v2-production-adoption/).) Even
though the Linkerd community [got pretty
good](https://linkerd.io/2016/06/17/small-memory-jvm-techniques-for-microservice-sidecars/)
at tuning the JVM's memory use to minimize footprint, it was still too much to
ask in a per-pod service mesh deployment model. So we knew we'd need a language
that compiled to native binaries, like Rust, Go, and C++.

Now, latency. Another lesson we learned from Linkerd 1.x that informed our
choice of Rust was the impact of garbage collection. In a garbage-collected
runtime, the GC must occasionally traverse the graph of objects in memory to
find the ones that are no longer in use and can be reclaimed. This process
takes time and can happen at unpredictable points. If a request comes in during
a garbage collector pass, it can have significantly heighted latency. This
spiky, unpredictable latency profile is the opposite of what we wanted for a
service mesh proxy. Thus, although [we love
Go](https://github.com/linkerd/linkerd2/search?l=go) (and the Linkerd 2.x

> control plane is written in it), Go, too, [is a garbage-collected
> language](https://github.com/linkerd/linkerd2/search?l=go). So that left us

Reviewer: this link looks wrong; i think i meant for it to point at 
```suggestion
control plane is written in it), Go, too, [is a garbage-collected
language](https://blog.discord.com/why-discord-is-switching-from-go-to-rust-a190bbca2b1f). So that left us
```

rather than pointing at all the linkerd Go code a second time (my bad).

[Mononoke](https://github.com/facebookexperimental/eden/blob/d374ea5098b554164a6210d8d0db89939be9eb44/eden/mononoke/Cargo.toml#L119),
and at companies like
[AWS](https://aws.amazon.com/blogs/opensource/aws-sponsorship-of-the-rust-project/)
to
[Discord](https://blog.discord.com/why-discord-is-switching-from-go-to-rust-a190bbca2b1f).

Hyper is Rust's leading asynchronous HTTP implementation, notable for its
[best-in-class
performance](https://www.techempower.com/benchmarks/#section=data-r18&hw=ph&test=plaintext)
and correctness. Like Tokio, Hyper has been battle-hardened by widespread use
at scale.

To secure meshed traffic with mutual TLS, the Linkerd proxy uses
[rustls](https://github.com/ctz/rustls), an implementation of the TLS protocol,
which is built on top of [_ring_](https://github.com/briansmith/ring) and
[webpki](https://github.com/briansmith/webpki), libraries that provide
underlying cryptographic primitives. An [independent security
audit](https://github.com/ctz/rustls/blob/master/audit/TLS-01-report.pdf),
sponsored by the CNCF, found that this cryptographic stack to be exceptionally
high quality, with the auditors from Cure53 “incredibly impressed with the

> presented software.”

Reviewer: should these "smart" quotes be replaced with "real" quotes?

else we can do.  Similarly, TCP traffic in an unknown protocol will be
transparently forwarded to its original destination.

On the other hand, what if the encrypted connection _is_ for us, as part of
Linkerd's [automatic mutual TLS](https://linkerd.io/2/features/automatic-mtls/)
feature? Each proxy in the mesh has its own unique cryptographic identity, the
key material for which is generated by the proxy on startup and never leaves
the pod boundary or gets written to disk. These identities are [signed by the
control plane's Identity
service](https://linkerd.io/2/features/automatic-mtls/#how-does-it-work) to
indicate that the proxy is authenticated to serve traffic for the [Kubernetes
ServiceAccount](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/)
of the pod that the proxy is injected into. If the SNI name matches the proxy's
service account, then we decrypt it and start processing it as part of the
service mesh.

Next, what does the proxy do if a request is meshed? Let's consider the case
where a meshed client sends an outbound request to its proxy. The proxy
performs the protocol detection we discussed above, and determines that this is

> a HTTP/1, HTTP/= 1, or gRPC request—protocols that Linkerd understands and can

Reviewer: ```suggestion
a HTTP/1, HTTP/2, or gRPC request—protocols that Linkerd understands and can
```

the application. In conjunction with Linkerd's intelligent, protocol-aware load
balancing, this is one of the reasons that meshed traffic can often have lower
latency than unmeshed traffic, despite taking additional network hops.

Putting it all together, the basic flow of logic in the proxy looks like this:

![](/uploads/flow-chart.png)

Although it provides a lot of functionality, we've kept Linkerd2-proxy as
simple and minimalist as possible. Best of all, the proxy's modular
architecture means that most features can be implemented as small,
self-contained modules and plugged in at the appropriate point in the stack,
keeping overall code complexity low.

## It's the proxy, silly

Today, Linkerd is the only service mesh to feature a data plane proxy designed
from the ground up explicitly for the service mesh use case. By focusing on the
service mesh's unique requirements and making full use of Rust's impressive
performance, security guarantees, and cutting-edge asynchronous networking
stack, we believe Linkerd2-proxy is the secret sauce to Linkerd's success.

{{< tweet 1091591745870139392 >}}
{{< tweet 1274302614231044096 >}}
{{< tweet 1278698855622721536 >}}

> So, do _you_ want to get involved in a cutting edge open source Rust project

Reviewer: love this
